src/freertos_drivers/esp32/Esp32WiFiManager.cxx

```diff
@@ -0,0 +1,985 @@
+    // the connection process, note it may not have an IP address immediately
+    // thus the need to check the connection result a few times before giving
+    // up with a FATAL error.
+    LOG(INFO, "[WiFi] Starting WiFi stack");
+    ESP_ERROR_CHECK(esp_wifi_start());
+
+    uint8_t attempt = 0;
+    EventBits_t bits;
+    uint32_t bitMask = WIFI_CONNECTED_BIT;
+    while(++attempt <= MAX_CONNECTION_CHECK_ATTEMPTS)
+    {
+        // If we have connected to the SSID we then are waiting for DHCP.
+        if(bits & WIFI_CONNECTED_BIT)
+        {
+            LOG(INFO, "[DHCP] [%d/%d] Waiting for IP address assignment.",
+                attempt, MAX_CONNECTION_CHECK_ATTEMPTS);
+        }
+        else
+        {
+            // Waiting for SSID connection
+            LOG(INFO, "[WiFi] [%d/%d] Waiting for SSID connection.",
+                attempt, MAX_CONNECTION_CHECK_ATTEMPTS);
+        }
+        bits = xEventGroupWaitBits(wifiStatusEventGroup_,
+            bitMask,        // bits we are interested in
+            pdFALSE,        // clear on exit
+            pdTRUE,         // wait for all bits
+            WIFI_CONNECT_CHECK_INTERVAL);
+        // Check if have connected to the SSID
+        if(bits & WIFI_CONNECTED_BIT)
+        {
+            // Since we have connected to the SSID we now need to track that we
+            // get an IP via DHCP.
+            bitMask |= DHCP_GOTIP_BIT;
+        }
+        // Check if we have received an IP via DHCP.
+        if(bits & DHCP_GOTIP_BIT)
+        {
+            break;
+        }
+    }
+
+    // Check if we successfully connected or not. If not, force a reboot.
+    if((bits & WIFI_CONNECTED_BIT) != WIFI_CONNECTED_BIT)
+    {
+        LOG(FATAL, "[WiFi] Failed to connect to SSID.");
+    }
+
+    // Check if we successfully connected or not. If not, force a reboot.
+    if((bits & DHCP_GOTIP_BIT) != DHCP_GOTIP_BIT)
+    {
+        LOG(FATAL, "[DHCP] Timeout waiting for an IP.");
+    }
+
+    // Initialize the mDNS system.
+    LOG(INFO, "[mDNS] Initializing mDNS system");
+    ESP_ERROR_CHECK(mdns_init());
+
+    // Set the mDNS hostname based on our generated hostname so it can be found
+    // by other nodes.
+    LOG(INFO, "[mDNS] Setting mDNS hostname to \"%s\"", hostname_.c_str());
+    ESP_ERROR_CHECK(mdns_hostname_set(hostname_.c_str()));
+
+    // Set the default mDNS instance name to the generated hostname.
+    ESP_ERROR_CHECK(mdns_instance_name_set(hostname_.c_str()));
+}
+
+// Starts a background task for the Esp32WiFiManager.
+void Esp32WiFiManager::start_wifi_task()
+{
+    LOG(INFO, "[WiFiMgr] Starting WiFi Manager task");
+    os_thread_create(&wifiTaskHandle_, "OpenMRN-WiFiMgr", WIFI_TASK_PRIORITY,
+        WIFI_TASK_STACK_SIZE, wifi_manager_task, this);
+}
+
+// Background task for the Esp32WiFiManager. This handles all outbound
+// connection attempts, configuration loading and making this node as a hub.
+void *Esp32WiFiManager::wifi_manager_task(void *param)
```

Comment:
this code is fine, it's idiomatic FreeRTOS code.

However, we are typically not writing low-level freertos code in OpenMRN, as it is not generally portable. Whenever possible we should try to use the abstractions in OS.hxx and Executor.
We have OSThread for infinite loops, StateFlow for sleeps-most-of-the-time use-cases, Notifiable and OSSem for wake-me-up-sometime callbacks, etc.

With moving towards WifiInterface it would be nice to make this code portable, because ultimately there is nothing freertos specific in here. In other words, all of this code should be able to run on a Raspberry Pi Zero W, or an ESP8266 NONOS, provided that an appropriate WifiInterface implementation is created.

This is clearly not a blocking comment, because we decided to take the WifiInterface refactoring to a separate effort.

Reply:
Yes, it does make sense to move this to a more generic approach using OSThread/etc. We just need a way to wake up the task when an "interesting" event occurs so it can take care of cleanup or start running, etc.

With the WiFiInterface code I think a common implementation can be achieved between CC32xx, ESP, RPi, ESP32 with a minimal implementation layer for the hardware drivers etc.